### ISSUE TITLE
stale brick process when volume stop operations are done in parallel

### ISSUE THREAD

**Issue by atinmu** [gluster/glusterd2#1470]
During stopBricks () , if we happen to send the volume stop requests in parallel, GD2 might fail to determine the last brick in the proc entry because -
Given it tries to match if for the given path if portmapper has only one entry, the sign out even might be processed later and even though it's the last brick to be detached function IsLastBrickInProc () will return > 1.

We should be moving towards a list based approach to store (in memory) all the bricks attached to the parent process and then build up the logic.

**Comment by vpandey-RH** [comment]
@atinmu I am having trouble understanding how a list might help. The portmapper we use currently is kind of an in-memory multi list itself.

**Comment by vpandey-RH** [comment]
@atinmu I have one idea which doesn't include adding a whole new data-structure. Instead of just having a mapping of port->brickpath->pid, can we add some sort of current state like pending-delete. So as soon as I send a detach request I can update the port mapper to change the state of brickpath to pending-delete and decrease the number of active bricks on a port(or when calculating IsLastBrickInProc() consider bricks with this state as well), when signOut req comes it can match pending-delete state and delete the brick entry.

**Comment by aravindavk** [comment]
> @atinmu @aravindavk I have one idea which doesn't include adding a whole new data-structure. Instead of just having a mapping of port->brickpath->pid, can we add some sort of current state like pending-delete. So as soon as I send a detach request I can update the port mapper to change the state of brickpath to pending-delete and decrease the number of active bricks on a port(or when calculating IsLastBrickInProc() consider bricks with this state as well), when signOut req comes it can match pending-delete state and delete the brick entry.

How about removing brickpath from portmap registry, signout coming after this can be safely discarded. Port collision will not happen since glusterd2 is not choosing the port.

**Comment by vpandey-RH** [comment]
@aravindavk Yes, I totally agree on this. We just need to think if this could have any negative effect in any use case, once we scale up.

**Comment by aravindavk** [comment]
only problem I could see is when glusterd2 restarts, this registry is reset. So if a Volume delete comes after the glusterd2 restart and before all bricks signed in, then the check will say that is the last brick and glusterd2 may terminate the process even though other bricks exists in that process.

**Comment by vpandey-RH** [comment]
That is one issue. Unless we handle signin in GD2 as well.

**Comment by atinmu** [comment]
Taking this out from GCS/1.0 tag considering we're not going to make brick multiplexing a default option in GCS/1.0 release.
